src/lib/operators/index_scan.cpp

```diff
@@ -0,0 +1,154 @@
+
+      // The output chunk is allocated on the same NUMA node as the input chunk. Also, the AccessCounter is
+      // reused to track accesses of the output chunk. Accesses of derived chunks are counted towards the
+      // original chunk.
+      Chunk chunk_out(chunk_guard->get_allocator(), chunk_guard->access_counter());
+
+      for (ColumnID column_id{0u}; column_id < _in_table->column_count(); ++column_id) {
+        auto ref_column_out = std::make_shared<ReferenceColumn>(_in_table, column_id, matches_out);
+        chunk_out.add_column(ref_column_out);
+      }
+
+      std::lock_guard<std::mutex> lock(output_mutex);
+      _out_table->emplace_chunk(std::move(chunk_out));
+    });
+
+    jobs.push_back(job_task);
+    job_task->schedule();
+  }
+
+  CurrentScheduler::wait_for_tasks(jobs);
+
+  return _out_table;
+}
+
+void IndexScan::_on_cleanup() {}
+
+void IndexScan::_validate_input() {
+  Assert(_scan_type != ScanType::OpLike, "Scan type not supported by index scan.");
+  Assert(_scan_type != ScanType::OpNotLike, "Scan type not supported by index scan.");
+
+  Assert(_left_column_ids.size() == _right_values.size(),
+         "Count mismatch: left column IDs and right values don’t have same size.");
+  if (_scan_type == ScanType::OpBetween) {
+    Assert(_left_column_ids.size() == _right_values2.size(),
+           "Count mismatch: left column IDs and right values don’t have same size.");
+  }
+
+  Assert(_in_table->get_type() == TableType::Data, "IndexScan does only support persistent tables right now.");
+}
+
+PosList IndexScan::_scan_chunk(const ChunkID chunk_id) {
+  const auto to_row_id = [chunk_id](ChunkOffset chunk_offset) { return RowID{chunk_id, chunk_offset}; };
+
+  auto range_begin = BaseIndex::Iterator{};
+  auto range_end = BaseIndex::Iterator{};
+
+  const auto& chunk = _in_table->get_chunk(chunk_id);
+  auto matches_out = PosList{};
+
+  const auto index = chunk.get_index_for(_index_type, _left_column_ids);
+  Assert(index != nullptr, "Index of specified type not found for column (vector).");
+
+  switch (_scan_type) {
+    case ScanType::OpEquals: {
+      range_begin = index->lower_bound(_right_values);
+      range_end = index->upper_bound(_right_values);
+      break;
+    }
+    case ScanType::OpNotEquals: {
+      // first, get all values less than the search value
+      range_begin = index->cbegin();
+      range_end = index->lower_bound(_right_values);
+
+      matches_out.reserve(std::distance(range_begin, range_end));
+      std::transform(range_begin, range_end, std::back_inserter(matches_out), to_row_id);
+
+      // set range for second half to all values greater than the search value
+      range_begin = index->upper_bound(_right_values);
+      range_end = index->cend();
+      break;
+    }
+    case ScanType::OpLessThan: {
+      range_begin = index->cbegin();
+      range_end = index->lower_bound(_right_values);
+      break;
+    }
+    case ScanType::OpGreaterThanEquals: {
```

Comment:
The order of cases here is a bit weird. I was expecting `OpLessThanEquals` here and was confused what you are doing